collector/src/main/java/io/prometheus/jmx/JmxCollector.java

```diff
@@ -335,9 +337,50 @@ private static boolean isLegalCharacter(char input) {
             (input >= '0' && input <= '9'));
   }
 
-    class Receiver implements JmxScraper.MBeanReceiver {
+    /**
+     * A sample is uniquely identified by its name, labelNames and labelValues
+     */
+    static class SampleKey {
+      String name;
+      List<String> labelNames;
+      List<String> labelValues;
+
+      SampleKey(String name, List<String> labelNames, List<String> labelValues) {
+        this.name = name;
+        this.labelNames = labelNames;
+        this.labelValues = labelValues;
+      }
+
+      static SampleKey of(MetricFamilySamples.Sample sample) {
+        return new SampleKey(sample.name, sample.labelNames, sample.labelValues);
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        SampleKey sampleKey = (SampleKey) o;
+
+        if (name != null ? !name.equals(sampleKey.name) : sampleKey.name != null) return false;
+        if (labelNames != null ? !labelNames.equals(sampleKey.labelNames) : sampleKey.labelNames != null) return false;
+        return labelValues != null ? labelValues.equals(sampleKey.labelValues) : sampleKey.labelValues == null;
```

Comment:
A little micro-optimization: We could compare `labelValues` first and then `labelNames`, because `labelValues` are more likely to be different.

Reply:
good call.